components/board/client.go

```diff
@@ -253,6 +265,125 @@ func (dic *digitalInterruptClient) RemoveCallback(c chan Tick) {
 	panic(errUnimplemented)
 }
 
+func (c *client) StreamTicks(ctx context.Context, interrupts []string, ch chan Tick, extra map[string]interface{}) error {
+	c.streamMu.Lock()
+	defer c.streamMu.Unlock()
+	ext, err := protoutils.StructToStructPb(extra)
+	if err != nil {
+		return err
+	}
+	c.extra = ext
+
+	// Start the stream
+	c.streamRunning = true
+	c.activeBackgroundWorkers.Add(1)
+	ctx, cancel := context.WithCancel(ctx)
+	c.cancelBackgroundWorkers = cancel
+	// Create a background go routine that connects to the server's stream
+	utils.PanicCapturingGo(func() {
+		defer c.activeBackgroundWorkers.Done()
+		c.connectTickStream(ctx, interrupts, ch)
+	})
+	c.mu.RLock()
+	ready := c.streamReady
+	c.mu.RUnlock()
+
+	// wait until the stream is ready to return
+	for !ready {
+		c.mu.RLock()
+		ready = c.streamReady
+		c.mu.RUnlock()
+		// wait for 50 ms before checking again
+		if !utils.SelectContextOrWait(ctx, 50*time.Millisecond) {
+			return ctx.Err()
+		}
+	}
+	return nil
+}
+
+func (c *client) connectTickStream(ctx context.Context, interrupts []string, ch chan Tick) {
+	defer func() {
+		c.streamMu.Lock()
+		defer c.streamMu.Unlock()
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		c.streamCancel = nil
+		c.streamRunning = false
+		c.streamReady = false
+	}()
+
+	streamCtx, cancel := context.WithCancel(ctx)
+	c.streamCancel = cancel
+
+	for {
+		c.mu.Lock()
+		c.streamReady = false
+		c.mu.Unlock()
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		req := &pb.StreamTicksRequest{
+			Name:     c.info.name,
+			PinNames: interrupts,
+		}
+
+		stream, err := c.client.StreamTicks(streamCtx, req)
+		if err != nil {
+			c.logger.CError(ctx, err)
+			if utils.SelectContextOrWait(ctx, 3*time.Second) {
+				continue
+			} else {
+				return
+			}
+		}
+
+		// repeatly receive from the stream
+		for {
+			select {
+			case <-ctx.Done():
+				c.logger.CError(ctx, err)
+				return
+			default:
+			}
+
+			c.mu.Lock()
+			c.streamReady = true
+			c.mu.Unlock()
+			streamResp, err := stream.Recv()
+			if err != nil && streamResp == nil {
+				// only debug log the context canceled error
+				c.logger.Debug(err)
+				c.streamCancel()
+				return
+			}
+			if err != nil {
+				c.logger.CError(ctx, err)
+			} else {
+				// If there is a response with a tick, send it to the channel
+				tick := Tick{
+					Name:             streamResp.PinName,
+					High:             streamResp.High,
+					TimestampNanosec: streamResp.Time,
+				}
+				ch <- tick
+			}
+		}
+	}
+}
+
+// Close cleanly closes the underlying connections.
+func (dic *digitalInterruptClient) Close(ctx context.Context) error {
+	if dic.cancelBackgroundWorkers != nil {
+		dic.cancelBackgroundWorkers()
```

Comment:
Given that `cancelBackgroundWorkers` can mutate, protect it with a mutex. Otherwise, you can get two goroutines both calling `Close()` at the same time. The first one will see that it's non-nil, the second one will see that it's non-nil, the first one will call it and set it to nil, and the second one will attempt to call it and panic.

Another possibility here is to change the design so that it's never mutated after construction and make sure things will work if it's called multiple times. This is particularly useful if something else might hold on to the mutex for a long time, thus making `Close()` potentially a very slow function.